test/cpp/end2end/xds_end2end_test.cc

```diff
@@ -4348,48 +4346,64 @@ TEST_P(LdsRdsTest, XdsRoutingApplyXdsTimeout) {
     balancers_[0]->ads_service()->SetLdsResource(listener);
   }
   // Test grpc_timeout_header_max of 1.5 seconds applied
-  auto t0 = system_clock::now();
+  gpr_timespec est_timeout_time = gpr_time_add(
+      gpr_now(GPR_CLOCK_MONOTONIC),
+      gpr_time_from_millis(
+          kTimeoutGrpcTimeoutHeaderMaxSecond * 1000 + kTimeoutMillis,
+          GPR_TIMESPAN));
+  gpr_timespec est_upperbound =
+      gpr_time_add(gpr_now(GPR_CLOCK_MONOTONIC),
+                   gpr_time_from_millis(
+                       kTimeoutMaxStreamDurationSecond * 1000 + kTimeoutMillis,
+                       GPR_TIMESPAN));
   CheckRpcSendFailure(1,
                       RpcOptions()
                           .set_rpc_service(SERVICE_ECHO1)
                           .set_rpc_method(METHOD_ECHO1)
                           .set_wait_for_ready(true)
                           .set_timeout_ms(kTimeoutApplicationSecond * 1000),
```

Comment:
`set_timeout_ms` will call `grpc_timeout_milliseconds_to_deadline` which measures `now` again and potentially uses `grpc_test_slowdown_factor`. This could be a potential problem when running it under sanitizers. (But this doesn't explain why it's done sooner)

Another angle I'm thinking of is type conversion between long and double happening a couple of times along the way. Not sure which one is problematic but it can introduce some error.